### PR TITLE
#81 frontend-is-not-your-enemy.md

### PR DESCRIPTION
**Pull request by jasofalcon**
There's a possibility it might look too long. In that case I would possibly remove an article or two and maybe leave them for part 2 or another article...

### CODE REVIEW

**Comment by abeggchr** [comment]
Did you already conduct a review with a peer? Is this article ready for the review of a lead author?

File: articles/frontend-is-not-your-enemy.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+For many years, developers are discriminating frontend technologies because of various reasons.
+
+*“It is too volatile!”*, some said
+
+*“But I've just learned AngularJS version 1.0 yesterday, today it's already v3.3.1”*, many yelled while hiding the tears...
+
+*“It doesn't work the same in every browser!”*, was heard from distance...
+
+We admit it. It is partially our fault. Us, engineers.
```
Comment (milanm) [inline articles/frontend-is-not-your-enemy.md]: I would shorten the whole intro story into 1-2 sentence, as you have more then allowed characters.

File: articles/frontend-is-not-your-enemy.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+However, a big portion of the problem are the users. It is similar in other industries. Look at the hardware vendors, trying to create just perfect user experience. Now I have to call my Mom every time I get a new washing machine because I don't understand any of the icons on it. Similarly, frontend has exploded in last decade or so, due to so many different (innovative) requirements to create the most user-friendly web app.
```
Comment (milanm) [inline articles/frontend-is-not-your-enemy.md]: I would rephrase this section, without this comparison to hardware vendors (and mommy :), just stick to the software industry, if possible.

File: articles/frontend-is-not-your-enemy.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+- Data binding
+- Routing
+- High performance templates
+- State management
+- Testing utilities
+
+At the time being, I would just highly recommend to write a small *todo* app in React, Angular and Vue.
+
+## Testing
```
Comment (milanm) [inline articles/frontend-is-not-your-enemy.md]: I would shorten this section a lot. You can just give in one or two sentences  these popular frameworks, without a description, as you don't have so much space.

File: articles/frontend-is-not-your-enemy.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+**Node** basically opened the door to a whole new universe of magic we can do in front-end development. And do so easily.
+
+It started with simple task-runners that can help you organize your code, clean-up, uglify/minify it, and even transpile it to something different.
+
+First tools worth mentioning was **Grunt** and **Gulp**, with much nicer syntax. Now we have **Webpack**, and just recently, version 4 was released. By the time you are reading this, it is probable that v5.0 is out. Also, there's a new Webpack nemesis called Parcel. It's in a hype phase now and many think it will overtake Webpack in upcoming years...
+
+Regardless of the tools you prefer, it is suggested to, when deciding on the tool, do thorough analysis, and choose the one which is battle tested, has a great community support, is feature-rich, and for complex project - is it performant.
+
+## Frameworks and Libraries
```
Comment (milanm) [inline articles/frontend-is-not-your-enemy.md]: Maybe to say here about difficulty to choose a framework in a large ocean of those.

File: articles/frontend-is-not-your-enemy.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+### Mocha
+
+Probably the most used one nowadays. It is very flexible, although you need some extra time to setup as you will probably want to install assertion or mocking libraries as well.
+
+### Jest
+
+Became rapidly popular due to the fact it came from Facebook alongside React. It plays well in benchmarkings due to its parallel testing. Jest also has a coolest logo and a good way to keep your UI tested using snapshots.
+
+## Final words
```
Comment (milanm) [inline articles/frontend-is-not-your-enemy.md]: If you don't have enough space, I would remove this section completely.

File: articles/frontend-is-not-your-enemy.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+```
+
+You are probably thinking: "Ok well const is ES6, my browser will tell me i cannot reassign it". That is correct, but now, you will be notified at compile-time, meaning your app won't build until you fix the issue. Thus, we have strictly typed JavaScript app.
+
+It has all the OO principles that Java has (and many much more sophisticated), such as interfaces, abstract classes, enums, and many many more.
+
+If this catches your attention, I advise to visit https://www.typescriptlang.org/docs/handbook/basic-types.html for more detailed information.
+
+## Build Tools and Node
```
Comment (milanm) [inline articles/frontend-is-not-your-enemy.md]: This section can be shorten, too. As you will still probably have 200-300 word over limit. Regarding that, maybe you can remove all the examples from the text and to leave just descriptions of concepts.

**Comment by jasofalcon** [comment]
Thanks for the hints Milan, I've adapted it and managed to bring the size down to ~750.

File: articles/frontend-is-not-your-enemy.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+Since frontend apps are becoming more and more complex and heavy with logic, it is crucial to treat them like grown-ups. Thus, testing JavaScript code is not a luxury any more, but rather a duty.
+Again, _"WHICH ONE?"_, you might scream, and again, _"It depends_", I shall respond.
+
+As you may guess, each framework has its own testing tool.
+These are some popular ones:
+
+### Jasmine
+
+One of the most popular testing frameworks. Simple to set-up and you can have your first tests in minutes. Widely respected by Angular lovers.
```
[the remark this answers is not in the document]
Comment (jasofalcon) [reply articles/frontend-is-not-your-enemy.md]: Ok, makes sense

File: articles/frontend-is-not-your-enemy.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+Most of mature ones provide similar solutions and follow kind of alike patterns:
+
+* Data binding
+* Routing
+* High performance templates
+* State management
+* Virtual DOM
+* Testing utilities
+
+_"But which one?"_ Well, it is indeed painful to choose a best framework in the ocean of great ones. Today, I would recommend writing a small _todo_ app in **React**, **Angular** and **Vue**.
```
[the remark this answers is not in the document]
Comment (jasofalcon) [reply articles/frontend-is-not-your-enemy.md]: Yeah, I'll use one of those two :+1:

File: articles/frontend-is-not-your-enemy.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+## Build Tools
+
+Yes, there's a proper build in frontend. For a while now. And hear this, one can even import .css in .js files now! I know, what kind of monster would do such a thing. Well...
+
+**Node** basically opened the door to a whole new universe of wizardry we can do in front-end development. Node is a popular JavaScript runtime built with V8 engine (same as Chrome). With Node, you can easily build scripts or web apis that execute on any OS environment. Many big players use it, such as Uber, Netflix, PayPal, etc... 
+It started with simple task-runner packages (Grunt and Gulp) that can help you organize your code, clean-up, uglify/minify it, and much more. Now we have **Webpack**, the most popular build and building tool on frontend. Using various loaders, you can easily import modules and do all kinds of transformations on your code, from development to deployment phase.
+
+Regardless of the tools you prefer, it is suggested to, when deciding on the tool, do thorough analysis, and choose the one which is battle tested, has a great community support, is feature-rich, and for complex project - is performant.
```
[the remark this answers is not in the document]
Comment (jasofalcon) [reply articles/frontend-is-not-your-enemy.md]: Will do :+1:

File: articles/frontend-is-not-your-enemy.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+I advise to look into ES6 documentation and get familiar with the features.
+
+## Compile Time Checking
+
+Ok, I know who you are... You are one of those people who call themselves _back-end_ developer. And you probably worked in strongly typed languages, and you are used to having your IDE insult you with **_Cannot assign boolean to String!_** So you find frontend rather repulsive.
+
+Well guess what, you have all that now with TypeScript! You can even customize the guidelines which the team will agree upon, and engineers will be notified if they for example define a method without a return type.
+
+You are probably thinking: _"Ok well keyword **const** exists in ES6, my browser will tell me I cannot reassign it, I don't need TypeScript"_. That is correct, but now, you will be notified at compile-time, meaning your app won't build until you fix the issue. Thus, we have strongly typed JavaScript app.
```
[the remark this answers is not in the document]
Comment (jasofalcon) [reply articles/frontend-is-not-your-enemy.md]: It should describe the difference between loading the page and seeing an error in browser console vs seeing it during 'compilation'. But it is not a major thing. I'll remove it

File: articles/frontend-is-not-your-enemy.md [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+Biggest leap forward - Classes! There is no longer need for that cold sweat before we use Prototype or function constructors, hoping chaos won't unfold in process.
+
+There are other features which you will fall in love with. Like _let, const, lambdas, destructuring, async/await, default params..._
+
+I advise to look into ES6 documentation and get familiar with the features.
+
+## Compile Time Checking
+
+Ok, I know who you are... You are one of those people who call themselves _back-end_ developer. And you probably worked in strongly typed languages, and you are used to having your IDE insult you with **_Cannot assign boolean to String!_** So you find frontend rather repulsive.
```
[the remark this answers is not in the document]
Comment (jasofalcon) [reply articles/frontend-is-not-your-enemy.md]: I see what you mean, Ok I'll try to rephrase it to make more sense for wider audience 👍

File: articles/frontend-is-not-your-enemy.md [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+* _“But I've just learned Angular 1.0 yesterday, today it's already v3.3.1”_, a man yelled while hiding the tears...
+
+However, frontend has exploded in last decade, due to so many innovative requirements to create the most user-friendly web app.
+Nonetheless, many good things have come out of this wormhole...
+
+## ECMAScript and OOP
+
+ECMAScript is a standard which drives all new native JavaScript features. ES**6** specifically, brought many improvements which developers craved for a long time.
+
+Biggest leap forward - Classes! There is no longer need for that cold sweat before we use Prototype or function constructors, hoping chaos won't unfold in process.
```
[the remark this answers is not in the document]
Comment (jasofalcon) [reply articles/frontend-is-not-your-enemy.md]: 👍 good hint

File: articles/frontend-is-not-your-enemy.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+# Frontend Is Not Your Enemy
+
+For many years, developers are discriminating frontend technologies.
+
+* _“It is too volatile!”_, some said
+
+* _“But I've just learned Angular 1.0 yesterday, today it's already v3.3.1”_, a man yelled while hiding the tears...
+
+However, frontend has exploded in last decade, due to so many innovative requirements to create the most user-friendly web app.
+Nonetheless, many good things have come out of this wormhole...
```
[the remark this answers is not in the document]
Comment (jasofalcon) [reply articles/frontend-is-not-your-enemy.md]: 👍

File: articles/frontend-is-not-your-enemy.md [hunk 14]
```diff
@@ -0,0 +1,87 @@
+---
+authorName: Janko Sokolović
+authorGithubUsername: jasofalcon
+issue: 81
+title: Frontend Is Not Your Enemy
+---
+# Frontend Is Not Your Enemy
+
+For many years, developers are discriminating frontend technologies.
```
[the remark this answers is not in the document]
Comment (jasofalcon) [reply articles/frontend-is-not-your-enemy.md]: I hope I understood the point clearly and I'll try to adapt it in a more reasonable way.
